fuse/dl/models/backbones/backbone_resnet_3d.py

```diff
@@ -17,78 +17,163 @@
+
+
+class BasicBlock(nn.Module):
+
+    expansion = 1
+
+    def __init__(
+        self,
+        inplanes: int,
+        planes: int,
+        conv_builder: Callable[..., nn.Module],
+        stride: int = 1,
+        downsample: Optional[nn.Module] = None,
+    ) -> None:
+        midplanes = (inplanes * planes * 3 * 3 * 3) // (inplanes * 3 * 3 + 3 * planes)
+
+        super().__init__()
+        self.conv1 = nn.Sequential(
+            conv_builder(inplanes, planes, midplanes, stride), nn.BatchNorm3d(planes), nn.ReLU(inplace=True)
+        )
+        self.conv2 = nn.Sequential(conv_builder(planes, planes, midplanes), nn.BatchNorm3d(planes))
+        self.relu = nn.ReLU(inplace=True)
+        self.downsample = downsample
+        self.stride = stride
+
+    def forward(self, x: Tensor) -> Tensor:
+        residual = x
+
+        out = self.conv1(x)
+        out = self.conv2(out)
+        if self.downsample is not None:
+            residual = self.downsample(x)
 
+        out += residual
+        out = self.relu(out)
 
-class BackboneResnet3D(VideoResNet):
+        return out
+
+
+class BasicStem(nn.Sequential):
+    """The default conv-batchnorm-relu stem"""
+
+    def __init__(self, in_channels=3, out_channels=64, kernel_size=(3, 7, 7), stride=(1, 2, 2)) -> None:
+        padding = tuple([x // 2 for x in kernel_size])
+        super().__init__(
+            nn.Conv3d(in_channels, out_channels, kernel_size=kernel_size, stride=stride, padding=padding, bias=False),
+            nn.BatchNorm3d(out_channels),
+            nn.ReLU(inplace=True),
+        )
+
+
+class BackboneResnet3D(nn.Module):
     """
-    3D model classifier (ResNet architecture"
+    A slightly more configureable ResNet3D implementation.
+    Default values are identical to the pytorch implementation.
     """
 
-    def __init__(self, pretrained: bool = False, in_channels: int = 3, name: str = "r3d_18", pool=False) -> None:
-        """
-        Create 3D ResNet model
-        :param pretrained: Use pretrained weights
-        :param in_channels: number of input channels
-        :param name: model name. currently only 'r3d_18' is supported
-        """
-        # init parameters per required backbone
-        init_parameters = {
-            "r3d_18": {
-                "block": BasicBlock,
-                "conv_makers": [Conv3DSimple] * 4,
-                "layers": [2, 2, 2, 2],
-                "stem": BasicStem,
-            },
-        }[name]
-        # init original model
-        super().__init__(**init_parameters)
-
-        # load pretrained parameters if required
+    def __init__(
+        self,
+        *,
+        in_channels: int = 3,
+        pool: bool = True,
+        layers: List[int] = [2, 2, 2, 2],
+        first_channel_dim: int = 64,
+        first_stride: int = 1,
+        stem_kernel_size: Sequence[int] = (3, 7, 7),
+        stem_stride: Sequence[int] = (1, 2, 2),
+        pretrained: bool = False,
+        name: str = "r3d_18",
```

Comment:
Can you document the important arguments like name and specify what the options are?

Reply:
done